packages/aws-client/src/S3.ts

```diff
@@ -342,34 +353,185 @@ export const s3PutObjectTagging = improveStackTrace(
+ *
+ * @example
+ * const object = await getObject({ Bucket: 'bucket', Key: 'key' });
+ * const object = await getObject({ Bucket: 'bucket', Key: 'key' }, { retries: 3 });
+ *
+ * @function
+ * @param {S3.GetObjectRequest} params - parameters expected by `S3.getObject()`
+ * @param {pRetry.Options} [retryOptions={ maxTimeout: 10_000, retries: 0 }] -
+ *    options to control retry behavior when an object does not exist (by
+ *    default, retries will not be performed)
+ * @returns {Promise} response from `S3.getObject()` as a promise
+ * @see {@link https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#getObject-property|S3.getObject}
+ * @see {@link https://github.com/tim-kos/node-retry#retryoperationoptions|retry options}
+ */
+const getObject = improveStackTrace(
+  retryable(
+    retryStatusCodes([304, 404, 412])(
+      (params: S3.GetObjectRequest) =>
+        s3().getObject(params).promise().then(checkMatchPreconditions(params))
     )
+  )
 );
 
+// Support for 2 function signatures for the `getS3Object` function.  The first
+// is to be backwards-compatible with the original signature that takes a bucket
+// and a key as distinct arguments.  The second is to support a full parameters
+// object to pass through directly to S3.getObject() so that we can pass in more
+// than only a bucket and a key.
+type S3GetObject = {
+  (
+    bucket: string,
+    key: string,
+    retryOptions?: pRetry.Options
+  ): S3RequestPromise<S3.GetObjectOutput>;
+  (
+    params: S3.GetObjectRequest,
+    retryOptions?: pRetry.Options
+  ): S3RequestPromise<S3.GetObjectOutput>;
+};
+
+/**
+ * Gets an object from S3, optionally retrying on failures (not found, not
+ * modified, or precondition failed).
+ *
+ * Supports specifying explicit bucket and key string arguments or a parameters
+ * argument (as expected by `S3.getObject()`), and optional retry options, in
+ * either case.
+ *
+ * @example
+ * getS3Object('bucket', 'key')
+ * getS3Object('bucket', 'key', { retries: 4 })
+ * getS3Object({ Bucket: 'bucket', Key: 'key' })
+ * getS3Object({ Bucket: 'bucket', Key: 'key', ...more })
+ * getS3Object({ Bucket: 'bucket', 'key' }, { retries: 3 })
+ *
+ * @function
+ * @param {string|S3.GetObjectRequest} bucketOrParams - name of bucket, or a
+ *    parameters object as expected by `S3.getObject`
+ * @param {string|pRetry.Options} keyOrRetryOptions - key for object
+ *    (filepath + filename), or retry options (see `retryOptions` parameter)
+ * @param {Object} [retryOptions={ maxTimeout: 10_000, retries: 0 }] - options
+ *    to control retry behavior when an object does not exist (by default,
+ *    retries will not be performed)
+ * @returns {Promise} response from `S3.getObject` as a promise
+ * @see {@link https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#getObject-property|S3.getObject}
+ * @see {@link https://github.com/tim-kos/node-retry#retryoperationoptions|retry options}
+ */
+export const getS3Object: S3GetObject = (
```

Comment:
So while I understand the intent here was to avoid a larger refactor and preserve the existing call signature of `getS3Object`, ultimately I feel like the maintainability burden of these changes is a bit too high and we've sacrificed some clarity/readability with them. 

I'd probably prefer to just add a `getS3ObjectWithEtag` function and leave `getS3Object` as it was. And remove all the other code in here that was added to support the refactor

Reply:
Yes, I'm not thrilled with the added complexity to the `getS3Object` function, but I'm also not thrilled with adding another function that does the _same_ thing, differing only in the expected parameters.

I did consider adding another function, but preferred not doing so, so if we do add a function, I want to make sure it can fully handle all parameters that we could possibly send to `S3.getObject`. If we add `getS3ObjectWithEtag`, and then later we want to pass yet another parameter through to S3, then we'll have to add yet another nearly duplicate function.

I was considering adding a `getObject` function that takes a `params` object to mirror the `S3.getObject` function that we're wrapping, but was concerned that having `getS3Object` _and_ `getObject` would be confusing (i.e., why 2 functions that do the same thing, but with different parameters?), unless we _also_ deprecate `getS3Object`.

What do you think about that?

Reply:
My vote would be to deprecate `getS3Object` and add something like this:

```ts
export const getObject = (
  params: AWS.S3.GetObjectRequest & { s3: AWS.S3 }
): Promise<AWS.S3.GetObjectOutput> =>
  params.s3.getObject(params).promise();
```